docs/internal/future_optimisations.md

```diff
@@ -0,0 +1,163 @@
+They are not in any particular order.
+
+Late selector colouring
+-----------------------
+
+In the MVP compiler, selector colouring is done in the front end.
+This means that we need to assign descriptor state even for interface calls that are possible to elide.
+The front end can make some early decisions:
+
+ - If only one concrete type is ever cast to an interface type then elide the interface entirely and reify everything for the concrete type.
+ - If a method on an interface type is never called, don't assign a selector for it.
+
+After inlining, some of these cases become clearer.
+For example, if a function that takes an interface and uses `match` expressions is inlined into the callers then it may be possible to resolve the matches entirely at compile time and completely eliminate the interface creation.
+This, in turn, means that there is no need to assign selectors for the methods that are used only by this interface (and no need to generate code for them if they are inlined).
+
+Late reification
+----------------
+
+In the MVP compiler, generics will be reified in the front end, methods that take interfaces will not be reified.
+After inlining the caller of a method that takes an interface parameter, we may discover that it is always called with the same concrete type at a specific call site.
+Reifying the called method at this point may make it a better inlining candidate if it eliminates some code in `match` expressions.
+
+For example, consider a function of the following shape:
+
+```verona
+someFunction(object : Interface)
+{
+	match (object)
+	{
+	  _MyPrivateConcreteType => fastPath(object);
+	   => slowPath(object;
+	}
+}
+```
+
+If this function is reified for a call site where the type is known to be `_MyPrivateConcreteType` then the reified version is a single tail call and so is trivially inlined.
+
+Conditional jumps for interface calls
+-------------------------------------
+
+Method invocations via interfaces are indirect jumps via the descriptor, roughly equivalent to the following in C:
+
+```c
+object->descriptor[selector](object);
+```
+
+For low-order polymorphic call sites, this can be more efficiently lowered to:
+
+```c
+if (object->descriptor == ClassA)
+	ClassAMethod(object);
+else if (object->descriptor == ClassB)
+	ClassBMethod(object);
+```
+
+This transform is often done by dynamic language VMs because it both plays well with branch predictors and because it exposes inlining opportunities.
+It may be beneficial to inline one or both of the methods, which improves locality of the code and then exposes further optimisation opportunities from specialising the method for this specific call site.
+
+Descriptor Elision
+------------------
+
+The descriptor is used for dynamic dispatch and for `match` expressions.
+If a class is never cast to an interface type (including a union type) then it should be possible to completely elide its descriptor.
```

Comment:
I thought this could be caught by the front end, but perhaps not through an interface, before inlining.

If that's the only case it's worth looking, then a comment here would be clearer.